feat: form components migration guides (#1501)
contentful/forma-36: MIGRATION.md

  <GridItem />
</Grid>;
```

## Form Components

> We changed how the Form Components work on the version 4, on the version 3 we had the Field components that would handle the everything related to a Field, like the Label, HelpText and ValidationMessage.  

Reviewer: ```suggestion
We changed how the Form Components work on version 4. In version 3 we had the Field components, like `TextField` or `SelectField` that would handle elements like Label, HelpText, and ValidationMessage.
```

> For the version 4 we created the `FormControl` which gives more flexibility to the user, when migrating to the version 4, you would need to use it.

Reviewer: ```suggestion
For version 4 we created the `FormControl` which gives more flexibility to the user and controls the given input, its validation, state, etc.
Within `FormControl` component, you have available all the previous options, so:
 - `FormControl.Label`
 - `FormControl.HelpText`
 - `FormControl.ValidationMessage` 
 - `FormControl.Counter`
```

> And all our Form Components can be used either as controlled or uncontrolled component, this is defined by using the `defaulValue`/`defaultChecked` prop for uncontrolled components and `value`/`isChecked` with an `onChange` props for the controlled version. For example:

Reviewer: ```suggestion
Any Forma 36 form element can be used either as controlled or uncontrolled components, by using the `defaulValue`/`defaultChecked` prop for uncontrolled and `value`/`isChecked` with an `onChange` props for the controlled version. For example:
```

```jsx static=true
// This would render the TextInput as an uncontrolled component
<TextInput defaultValue="Some initial Value" />

// This would render the TextInput as a controlled component
<TextInput value="Some initial Value" onChange={onChange} />
```

### Field Components

> The `CheckboxField`, `RadioButtonField`, `SelectField`, and `TextField` components were removed on the version 4, and need to be replaced by the new `FormControl` component and it's compound components. For example:

Reviewer: ```suggestion
The `CheckboxField`, `RadioButtonField`, `SelectField`, and `TextField` components were removed in version 4, and need to be replaced by the new `FormControl` component and its compound components. For example:
```

```tsx static=true
import { CheckboxField, TextField } from '@contentful/forma-36-react-components';

<CheckboxField
  labelText="Your label text"
  id="some-id"
  type="checkbox"
  helpText="Some help text"
  validationMessage="validation message"
/>

// TextInput
<TextField
  id="inputId"
  name="email"
  labelText="Counting characters and HelpText"
  textInputProps={{
    maxLength: 10,
  }}
  helpText="Some help text"
  validationMessage="Validation Message"
/>
```

into this new version:

```tsx static=true
import { Checkbox, TextInput, FormControl } from '@contentful/f36-components';

<FormControl id="some id">
  <Checkbox helpText="Some help text">Your label text</Checkbox>
  <FormControl.ValidationMessage>validation message</FormControl.ValidationMessage>
</FormControl>

<FormControl id="inputId">
  <FormControl.Label>Counting characters and HelpText</FormControl.Label>
  <TextInput name="email" maxLength={10} />
  <Flex justifyContent="space-between">
    <FormControl.HelpText>Some help text</FormControl.HelpText>
    <FormControl.Counter />
  </Flex>
  <FormControl.ValidationMessage>Validation Message</FormControl.ValidationMessage>
</FormControl>
```

### Checkbox

> With the removal of the `CheckboxField` component, we updated how this component works on the version 4, on the version 3 it only rendered the checkbox input without anything else, now it also handles the label and help text for the checkbox. For example:

Reviewer: ```suggestion
We removed `CheckboxField` component in version 4. Now instead of doing:
```

```jsx static=true
import { Checkbox } from '@contentful/forma-36-react-components';

<Checkbox id="checkbox" name="checkbox" labelText="some label" />;
```

> This example would render only the input and the `labelText` would be set as the `aria-label`.

Reviewer: Maybe something like:
Note: This example would render only the input and the `labelText` would be set as the `aria-label`.

> On the version 4:

Reviewer: ```suggestion
You will do:
```

```jsx static=true
import { Checkbox } from '@contentful/f36-components';

// This would render the exact same as on version 3
<Checkbox id="checkbox" name="checkbox" aria-label="some label" />

// You can also pass the label and help text to the checkbox
<Checkbox id="checkbox" name="checkbox" helpText="Some help text">
  This label will be rendered at the side of the checkbox
</Checkbox>
```

> We also had some API changes on the checkbox, check the props that were renamed:

Reviewer: ```suggestion
We also introduced API changes for the checkbox. This is an overview of the renamed properties:
```

- `required` was renamed to `isRequired`;
- `labelText` should be replaced by `aria-label` when not passing a label to the checkbox;
- `checked` was renamed to `isChecked`;
- `disabled` was renamed to `isDisabled`;
- `indeterminate` was renamed to `isIndeterminate`.

#### How to migrate your checkbox components

> The migration for the Checkbox needs to be manually as it depends on the context that the component is being used, for example:

Reviewer: ```suggestion
The migration for the version 3 `Checkbox` component needs to be manual as it depends on the context that the component is being used, for example:
```

```jsx static=true
import {
  Checkbox,
  Box,
  FormLabel,
} from '@contentful/forma-36-react-components';

<Box>
  <Checkbox id="checkbox" name="checkbox" />
  <FormLabel htmlFor="checkbox">Some label here</FormLabel>
</Box>;
```

> could become:

Reviewer: ```suggestion
becomes:
```

```jsx static=true
import { Checkbox } from '@contentful/f36-components';

<Checkbox id="checkbox" name="checkbox">
  Some label here
</Checkbox>;
```

If you only use `CheckboxField` component, please check the [Field components](#field-components) section.

### RadioButton

> We renamed the `RadioButton` component to be only `Radio`, also with the removal of the `RadioButtonField` component, we updated how this component works on the version 4, on the version 3 it only rendered the radio input without anything else, now it also handles the label and help text for it. For example:

Reviewer: ```suggestion
In the version 4 we renamed the `RadioButton` to `Radio`, and we removed `RadioButtonField` component. Now instead of doing:
```

```jsx static=true
import { RadioButton } from '@contentful/forma-36-react-components';

<RadioButton name="someOption" id="radio" labelText="some label" />;
```

> This example would render only the input and the `labelText` would be set as the `aria-label`.

Reviewer: ```suggestion
_Note: This example would render only the input and the `labelText` would be set as the `aria-label`._
```

> On the version 4:

Reviewer: ```suggestion
You will do the following:
```

```jsx static=true
import { Radio } from '@contentful/f36-components';

// This would render the exact same as on version 3
<Radio name="someOption" id="radio" aria-label="some label" />

// You can also pass the label and help text to the radio
<Radio name="someOption" id="radio" helpText="Some help text">
  This label will be rendered at the side of the input
</Radio>
```

> We also had some API changes on the radio props, check the props that were renamed:

Reviewer: ```suggestion
We also introduced API changes for the radio. This is an overview of the renamed properties:
```

- `required` was renamed to `isRequired`;
- `labelText` should be replaced by `aria-label` when not passing a label to the input;
- `checked` was renamed to `isChecked`;
- `disabled` was renamed to `isDisabled`.

#### How to migrate your RadioButton components

Run the [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod) to migrate your v3 `RadioButton` component to the new version:

`npx @contentful/f36-codemod`

If you decide to do it manually, you would need to transform your code:

```jsx static=true
import { RadioButton } from '@contentful/forma-36-react-components';

<RadioButton
  id="radio"
  disabled={false}
  checked={true}
  labelText="some label"
/>;
```

into this new version:

```jsx static=true
import { Radio } from '@contentful/f36-components';

<Radio
  id="radio"
  isDisabled={false}
  isChecked={true}
  aria-label="some label"
/>;
```

### Select

> We changed the `Select` component to be easier to use, instead of having to import both `Select` and `Option` components, on the version 4 we have the `Select.Option` compound component. We also had some API changes. For example:

Reviewer: ```suggestion
`Select` component becomes compound component in version 4. You can use `Select` and `Select.Option` , with just importing `Select`. You would need to transform your existing code:
```

```tsx static=true
<Select id="optionSelect" name="optionSelect" width="large">
  <Option value="optionOne">Option 1</Option>
  <Option value="optionTwo">Long Option 2</Option>
</Select>
```

> will become:

Reviewer: ```suggestion
into the following:
```

```tsx static=true
<Select id="optionSelect" name="optionSelect">
  <Select.Option value="optionOne">Option 1</Select.Option>
  <Select.Option value="optionTwo">Long Option 2</Select.Option>
</Select>
```

> We also had some API changes on the select, check the props that had changes:

Reviewer: ```suggestion
We also introduced some API changes for the select, check the props that had changes:
```

- `hasError` was renamed to `isInvalid`;
- `disabled` was renamed to `isDisabled`;
- The `Select.Option` component can receive a `isDisabled` prop instead of the `disabled`.

#### How to migrate your Select components

Run the [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod) to migrate your v3 `Select` component to the new version:

`npx @contentful/f36-codemod`

If you decide to do it manually, you would need to transform your code:

```jsx static=true
import { Select, Option } from '@contentful/forma-36-react-components';

<Select name="optionSelect" id="optionSelect">
  <Option value="optionOne">Option One</Option>
  <Option value="optionTwo">Option Two</Option>
  <Option value="optionThree" disabled>
    Option Three
  </Option>
</Select>;
```

into this new version:

```jsx static=true
import { Select } from '@contentful/f36-components';

<Select name="optionSelect" id="optionSelect">
  <Select.Option value="optionOne">Option One</Select.Option>
  <Select.Option value="optionTwo">Option Two</Select.Option>
  <Select.Option value="optionThree" isDisabled>
    Option Three
  </Select.Option>
</Select>;
```

### Switch

> In v4 the Switch component has received API improvements. The changes are based on our code style guide, which create consistent, easy to use APIs. For example:

Reviewer: ```suggestion
In v4 the `Switch` component has received API improvements, that are based on our code style guide, which creates consistent, easy-to-use APIs. For example:
```

```jsx static=true
import { Switch } from '@contentful/forma-36-react-components';

<Switch name="switch" id="switch" onToggle={onToggle} labelText="some label" />;
```

will become:

```jsx static=true
import { Switch } from '@contentful/f36-components';

<Switch name="switch" id="switch" onChange={onChange}>
  some label
</Switch>;
```

> When changing `onToggle` to `onChange`, before it would expect a function that receives boolean value, now it works as any other `onChange` on any form component, so the handling of on or off should be done by the user.

Reviewer: hmm, I don't understand this sentence. Maybe we could rewrite it a bit?

Author: Maybe something like this would make more sense?? It's basically because the onToggle received a boolean, and now the onChange receive the entire event instead. 🤔 
```suggestion
_Note: on v3 the `onToggle` method had as parameter the checked value (true/false), the `onChange` has the event as parameter, `event.target.checked` will have the true/false value_```

Reviewer: _Note: in v3 the `onToggle` method had `checked` value (true/false) as parameter, now the `onChange` method has the event as parameter, so `event.target.checked` will have the true/false value_

Maybe something lke that?

  aria-label="userEmail"
  className="my-extra-class"
  defaultValue="some value"
/>;

<TextInput
  id="someInput"
  name="userEmail"
  aria-label="userEmail"
  isDisabled
  isRequired
  value="some value"
  onChange={() => {}}
/>;

<TextInput
  id="someInput"
  name="userEmail"
  aria-label="userEmail"
  isInvalid
  isReadOnly
  onChange={() => {}}
/>;

<TextInput.Group>
  <TextInput id="someInput" name="userEmail" aria-label="userEmail" />
  <CopyButton />
</TextInput.Group>;

<Textarea
  id="someInput"
  name="userEmail"
  className="my-extra-class"
  isDisabled
  defaultValue="some value"
/>;

<Textarea
  id="someInput"
  name="userEmail"
  className="my-extra-class"
  isInvalid
  value="some value"
  onChange={() => {}}
/>;
```

You can also make these changes manually.

### FieldGroup

The `FieldGroup` component was removed, and should be replaced by one of the Layout components which best fits your need. For example:

```tsx static=true
<FieldGroup>...</FieldGroup>
```

for keeping the same spacing as before it should become:

```tsx static=true
<Flex flexDirection="column" gap="spacingXs">
  ...
</Flex>
```

#### How to migrate your FieldGroup components

> The migration should be done manually, using the Layout component that best fits the user needs, the layout components are `Box`, `Flex`, `Stack`, and `Grid`.

Reviewer: maybe links to those components might be useful.

Author: should we link to the repository or the documentation website?? I considered the latter, but I think we don't have the final URL for the v4 website

### Form

> The only thing that has changed on the Form component is that it doesn't have the spacing prop anymore. For example:

Reviewer: ```suggestion
In version 4 we removed `spacing` property for the Form component and that's the only API change for this component. For example:
```

```tsx static=true
import { Form } from '@contentful/forma-36-react-components';

<Form onSubmit={handleSubmit} spacing="condensed">
  ...
</Form>;
```

will become:

```tsx static=true
import { Form } from '@contentful/f36-components';

<Form onSubmit={handleSubmit} spacing="condensed">
  ...
</Form>;
```

#### How to migrate your form components

To migrate your v3 `Form` component to the version 4 you can simply run the [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod) that we prepared for this occasion, like:

`npx @contentful/f36-codemod`

What codemod will do for you, and what you can do yourself manually if you decide to do so, is:

it transforms:

```tsx static=true
import { Form } from '@contentful/forma-36-react-components';

<Form onSubmit={handleSubmit} spacing="condensed">
  Form elements goes here and spacing value is removed
</Form>;
```

into this new version:

```tsx static=true
import { Form } from '@contentful/f36-components';

<Form onSubmit={handleSubmit}>
  Form elements goes here and spacing value is removed
</Form>;
```

### FormLabel

> The FormLabel component was updated to be a compound component of the new FormControl component and it should be update to be wrapped by the FormControl component. For example:

Reviewer: ```suggestion
The `FormLabel` component was updated to be a compound component of the new `FormControl` component and it should be update to be wrapped by the `FormControl` component. For example:
```

Reviewer: Maybe we could link `FormControl` section in here? WDYT?

```tsx static=true
// Checkbox
<CheckboxField
  labelText="Your label text"
  id="some-id"
  type="checkbox"
  helpText="Some help text"
  validationMessage="validation message"
/>

// TextInput
<TextField
  id="inputId"
  name="email"
  labelText="Counting characters and HelpText"
  textInputProps={{
    maxLength: 10,
  }}
  helpText="Some help text"
  validationMessage="Validation Message"
/>
```

will become:

```tsx static=true
// Checkbox
<FormControl id="some-id">
  <Checkbox helpText="Some help text">Your label text</Checkbox>
  <FormControl.ValidationMessage>validation message</FormControl.ValidationMessage>
</FormControl>

// TextInput
<FormControl id="inputId">
  <FormControl.Label>Counting characters and HelpText</FormControl.Label>
  <TextInput name="email" maxLength={10} />
  <Flex justifyContent="space-between">
    <FormControl.HelpText>Some help text</FormControl.HelpText>
    <FormControl.Counter />
  </Flex>
  <FormControl.ValidationMessage>Validation Message</FormControl.ValidationMessage>
</FormControl>
```

> For more detailed information and examples, check our documentation website.

Author: @burakukula Do we know how the URL of the documentation will be?

</FormControl>
```

### Checkbox

We removed `CheckboxField` component in version 4. Now instead of doing:

```jsx static=true
import { Checkbox } from '@contentful/forma-36-react-components';

<Checkbox id="checkbox" name="checkbox" labelText="some label" />;
```

_Note: This example would render only the input and the `labelText` would be set as the `aria-label`._

You will do:

```jsx static=true
import { Checkbox } from '@contentful/f36-components';

// This would render the exact same as on version 3
<Checkbox id="checkbox" name="checkbox" aria-label="some label" />

// You can also pass the label and help text to the checkbox
<Checkbox id="checkbox" name="checkbox" helpText="Some help text">
  This label will be rendered at the side of the checkbox
</Checkbox>
```

We also introduced API changes for the checkbox. This is an overview of the renamed properties:

- `required` was renamed to `isRequired`;
- `labelText` should be replaced by `aria-label` when not passing a label to the checkbox;
- `checked` was renamed to `isChecked`;
- `disabled` was renamed to `isDisabled`;
- `indeterminate` was renamed to `isIndeterminate`.

#### How to migrate your checkbox components

> The migration for the version 3 `Checkbox` component needs to be manual as it depends on the context that the component is being used, for example:

Reviewer 2: You must manually migrate the version 3 `Checkbox` component as it depends on the context the component is being used. For example:

```jsx static=true
import {
  Checkbox,
  Box,
  FormLabel,
} from '@contentful/forma-36-react-components';

<Box>
  <Checkbox id="checkbox" name="checkbox" />
  <FormLabel htmlFor="checkbox">Some label here</FormLabel>
</Box>;
```

becomes:

```jsx static=true
import { Checkbox } from '@contentful/f36-components';

<Checkbox id="checkbox" name="checkbox">
  Some label here
</Checkbox>;
```

If you only use `CheckboxField` component, please check the [Field components](#field-components) section.

### RadioButton

> In the version 4 we renamed the `RadioButton` to `Radio`, and we removed `RadioButtonField` component. Now instead of doing:

Reviewer 2: In version 4 we renamed the `RadioButton` to `Radio` and we removed the `RadioButtonField` component. Now instead of:

```jsx static=true
import { RadioButton } from '@contentful/forma-36-react-components';

<RadioButton name="someOption" id="radio" labelText="some label" />;
```

_Note: This example would render only the input and the `labelText` would be set as the `aria-label`._

> You will do the following:

Reviewer 2: you can use:

```jsx static=true
import { Radio } from '@contentful/f36-components';

// This would render the exact same as on version 3
<Radio name="someOption" id="radio" aria-label="some label" />

// You can also pass the label and help text to the radio
<Radio name="someOption" id="radio" helpText="Some help text">
  This label will be rendered at the side of the input
</Radio>
```

We also introduced API changes for the radio. This is an overview of the renamed properties:

- `required` was renamed to `isRequired`;
- `labelText` should be replaced by `aria-label` when not passing a label to the input;
- `checked` was renamed to `isChecked`;
- `disabled` was renamed to `isDisabled`.

#### How to migrate your RadioButton components

> Run the [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod) to migrate your v3 `RadioButton` component to the new version:

Reviewer 2: To migrate your `RadioButton` component to v4, run the following [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod):

`npx @contentful/f36-codemod`

> If you decide to do it manually, you would need to transform your code:

Reviewer 2: If you want to do it manually, you must transform your code as follows:

```jsx static=true
import { RadioButton } from '@contentful/forma-36-react-components';

<RadioButton
  id="radio"
  disabled={false}
  checked={true}
  labelText="some label"
/>;
```

into this new version:

```jsx static=true
import { Radio } from '@contentful/f36-components';

<Radio
  id="radio"
  isDisabled={false}
  isChecked={true}
  aria-label="some label"
/>;
```

### Select

> `Select` component becomes compound component in version 4. You can use `Select` and `Select.Option` , with just importing `Select`. You would need to transform your existing code:

Reviewer 2: The `Select` component becomes a compound component in version 4. You can use `Select` and `Select.Option` by just importing `Select`. You must transform your existing code as follows:

```tsx static=true
<Select id="optionSelect" name="optionSelect" width="large">
  <Option value="optionOne">Option 1</Option>
  <Option value="optionTwo">Long Option 2</Option>
</Select>
```

> into the following:

Reviewer 2: into:

```tsx static=true
<Select id="optionSelect" name="optionSelect">
  <Select.Option value="optionOne">Option 1</Select.Option>
  <Select.Option value="optionTwo">Long Option 2</Select.Option>
</Select>
```

> We also introduced some API changes for the select, check the props that had changes:

Reviewer 2: We also introduced some API changes for the select. This is an overview of the changed props:

- `width` was removed;
- `required` was renamed to `isRequired`;
- `hasError` was renamed to `isInvalid`;
- `disabled` was renamed to `isDisabled`;
- The `Select.Option` component can receive a `isDisabled` prop instead of the `disabled`.

#### How to migrate your Select components

> Run the [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod) to migrate your v3 `Select` component to the new version:

Reviewer 2: To migrate your v3 `Select` component, run the following [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod):

`npx @contentful/f36-codemod`

> If you decide to do it manually, you would need to transform your code:

Reviewer 2: If you want to do it manually, you must transform your code as follows:

```jsx static=true
import { Select, Option } from '@contentful/forma-36-react-components';

<Select name="optionSelect" id="optionSelect">
  <Option value="optionOne">Option One</Option>
  <Option value="optionTwo">Option Two</Option>
  <Option value="optionThree" disabled>
    Option Three
  </Option>
</Select>;
```

into this new version:

```jsx static=true
import { Select } from '@contentful/f36-components';

<Select name="optionSelect" id="optionSelect">
  <Select.Option value="optionOne">Option One</Select.Option>
  <Select.Option value="optionTwo">Option Two</Select.Option>
  <Select.Option value="optionThree" isDisabled>
    Option Three
  </Select.Option>
</Select>;
```

### Switch

> In v4 the `Switch` component has received API improvements, that are based on our code style guide, which creates consistent, easy-to-use APIs. For example:

Reviewer 2: In v4 the `Switch` component has received API improvements. The changes are based on our code style guide, which create consistent, easy to use APIs. For example:

```jsx static=true
import { Switch } from '@contentful/forma-36-react-components';

<Switch name="switch" id="switch" onToggle={onToggle} labelText="some label" />;
```

> will become:

Reviewer 2: becomes:

```jsx static=true
import { Switch } from '@contentful/f36-components';

<Switch name="switch" id="switch" onChange={onChange}>
  some label
</Switch>;
```

When changing `onToggle` to `onChange`, before it would expect a function that receives boolean value, now it works as any other `onChange` on any form component, so the handling of on or off should be done by the user.

#### How to migrate your Switch components

> The migration for the Switch needs to be done manually by updating the usage of the props, for example:

Reviewer 2: You must manually migrate the Switch component by updating the props usage. For example:

  onToggle={onToggle}
  labelText="some label"
/>;
```

into this new version:

```jsx static=true
import { Switch } from '@contentful/f36-components';

const [isChecked, setIsChecked] = React.useState(false);
const onChange = (event) => setIsChecked(event.target.checked);
<Switch name="switch" id="switch" isChecked={isChecked} onChange={onChange}>
  some label
</Switch>;
```

### TextInput and TextArea

In v4 the TextInput and TextArea components have received API improvements. The changes are based on our code style guide, which create consistent, easy to use APIs. For example:

```jsx static=true
<TextInput
  id="someInput"
  name="userEmail"
  labelText="userEmail"
  className="my-extra-class"
  value="some value"
/>

<Textarea
  id="someInput"
  name="userEmail"
  className="my-extra-class"
  disabled
  value="some value"
/>;
```

becomes:

```jsx static=true
<TextInput
  id="someInput"
  name="userEmail"
  aria-label="userEmail"
  className="my-extra-class"
  defaultValue="some value" />

<Textarea
  id="someInput"
  name="userEmail"
  className="my-extra-class"
  isDisabled
  defaultValue="some value" />;
```

#### How to migrate your TextInput and TextArea components

> To migrate the `TextInput` or `TextArea` components to the v4 run the following [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod):

Reviewer 2: Remove the article before v4.

  isDisabled
  isRequired
  value="some value"
  onChange={() => {}}
/>;

<TextInput
  id="someInput"
  name="userEmail"
  aria-label="userEmail"
  isInvalid
  isReadOnly
  onChange={() => {}}
/>;

<TextInput.Group>
  <TextInput id="someInput" name="userEmail" aria-label="userEmail" />
  <CopyButton />
</TextInput.Group>;

<Textarea
  id="someInput"
  name="userEmail"
  className="my-extra-class"
  isDisabled
  defaultValue="some value"
/>;

<Textarea
  id="someInput"
  name="userEmail"
  className="my-extra-class"
  isInvalid
  value="some value"
  onChange={() => {}}
/>;
```

You can also make these changes manually.

### FieldGroup

The `FieldGroup` component was removed, and should be replaced by one of the Layout components which best fits your need. For example:

```tsx static=true
<FieldGroup>...</FieldGroup>
```

for keeping the same spacing as before it should become:

```tsx static=true
<Flex flexDirection="column" gap="spacingXs">
  ...
</Flex>
```

#### How to migrate your FieldGroup components

The migration should be done manually, using the Layout component that best fits the user needs, the layout components are `Box`, `Flex`, `Stack`, and `Grid`.

### Form

> In version 4 we removed `spacing` property for the Form component and that's the only API change for this component. For example:

Reviewer 2: In version 4, we improved the Form component API by removing the `spacing` property. For example:

```tsx static=true
import { Form } from '@contentful/forma-36-react-components';

<Form onSubmit={handleSubmit} spacing="condensed">
  ...
</Form>;
```

> will become:

Reviewer 2: becomes:

```tsx static=true
import { Form } from '@contentful/f36-components';

<Form onSubmit={handleSubmit} spacing="condensed">
  ...
</Form>;
```

> #### How to migrate your form components

Reviewer 2: How to migrate your Form components

> To migrate your v3 `Form` component to the version 4 you can simply run the [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod) that we prepared for this occasion, like:

Reviewer 2: To migrate your `Form` component to v4, run the following [codemod](https://github.com/contentful/forma-36/tree/forma-v4/packages/forma-36-codemod):

`npx @contentful/f36-codemod`

> What codemod will do for you, and what you can do yourself manually if you decide to do so, is:

Reviewer 2: When running the codemod, the following changes occur:

> it transforms:

Reviewer 2: We should remove this.

```tsx static=true
import { Form } from '@contentful/forma-36-react-components';

<Form onSubmit={handleSubmit} spacing="condensed">
  Form elements goes here and spacing value is removed
</Form>;
```

> into this new version:

Reviewer 2: becomes:

```tsx static=true
import { Form } from '@contentful/f36-components';

<Form onSubmit={handleSubmit}>
  Form elements goes here and spacing value is removed
</Form>;
```

### FormLabel

> The `FormLabel` component was updated to be a compound component of the new `FormControl` component and it should be update to be wrapped by the `FormControl` component. For example:

Reviewer 2: `FormLabel` becomes a compound component of the new `FormControl` component. It must be updated to be wrapped by the `FormControl` component. For example:

```jsx static=true
<FormLabel htmlFor="inputId" required>Label for the input</FormLabel>
<TextInput name="input" id="inputId">
```

becomes:

```jsx static=true
<FormControl.FormLabel htmlFor="inputId" isRequired>Label for the input</FormControl.FormLabel>
<TextInput name="input" id="inputId">

// You can also use the FormControl to wrap it, and it would handle the id and isRequired
<FormControl isRequired id="inputId">
  <FormControl.FormLabel>Label for the input</FormControl.FormLabel>
  <TextInput name="input">
</FormControl>
```

#### How to migrate your FormLabel components

> The migration for the FormLabel needs to be done manually by updating the usage. For example:

Reviewer 2: You must manually migrate the `FormLabel` component by updating the usage. For example:
